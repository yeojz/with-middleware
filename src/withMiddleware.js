import { createFactory } from 'react';
import applyMiddleware from './helpers/applyMiddleware';

function createComponent(stateName, dispatchName, createDispatch, factory) {
  function ComponentWithMiddleware(props) {
    const dispatch = createDispatch(props[stateName], props[dispatchName]);

    return factory(
      Object.assign({}, props, {
        [dispatchName]: dispatch
      })
    );
  }

  return ComponentWithMiddleware;
}

function withMiddleware(stateName, dispatchName, middlewares) {
  if (!Array.isArray(middlewares)) {
    throw new Error(
      `Expecting argument 3 of "withMiddleware" to be an array. ${typeof middlewares} given.`
    );
  }

  const createDispatch = applyMiddleware(...middlewares);

  return BaseComponent => {
    const factory = createFactory(BaseComponent);
    return createComponent(stateName, dispatchName, createDispatch, factory);
  };
}

export default withMiddleware;
