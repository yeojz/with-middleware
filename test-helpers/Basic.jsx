/* eslint-disable react/prop-types */

import React from 'react';

function Basic(props) {
  return (
    <div>
      <div data-testid="value">{props.counter}</div>
      <button
        data-testid="increment"
        onClick={() => props.setCounter(n => n + 1)}
      >
        Increment
      </button>
      <button
        data-testid="decrement"
        onClick={() => props.setCounter(n => n - 1)}
      >
        Decrement
      </button>
    </div>
  );
}

export default Basic;
