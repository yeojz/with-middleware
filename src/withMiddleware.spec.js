import React from 'react';
import withMiddleware from './withMiddleware';
import { compose, withState } from 'recompose';
import TestRenderer from 'react-test-renderer';
import Basic from '../tests/helpers/Basic';

describe('withMiddleware', () => {
  describe('Basic', () => {
    let called;
    const middleware = () => next => action => {
      const result = next(action);
      called(result);
      return result;
    };

    beforeEach(() => {
      called = jest.fn();
    });

    const enhancer = compose(
      withState('counter', 'setCounter', 0),
      withMiddleware('counter', 'setCounter', [middleware])
    );

    test('should render without errors', () => {
      const Component = enhancer(Basic);
      const result = TestRenderer.create(<Component />);
      expect(result).toMatchSnapshot();
    });

    test('should throw an error when invalid argument', () => {
      expect(() => withMiddleware('counter', 'setState')).toThrowError();
    });
  });
});
