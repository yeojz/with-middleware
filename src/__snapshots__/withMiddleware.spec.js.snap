// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withMiddleware Basic should render without errors 1`] = `
<div>
  <div
    className="value"
  >
    0
  </div>
  <button
    className="increment"
    onClick={[Function]}
  >
    Increment
  </button>
  <button
    className="decretment"
    onClick={[Function]}
  >
    Decrement
  </button>
</div>
`;
